Marketplace search view more button & categories

kdevnel opened the pull request:
### Changes proposed in this Pull Request:

Closes https://github.com/Automattic/woocommerce.com/issues/18213

- Add "View All" button to themes and extension on search page
- When clicked, the button will load only those results with the category selector shown
- Fixes some bugs with URL query parameters not fully being reset when switching tabs/pages
- Consolidates the `Extensions` and `Themes` components into a single `Products` component
- Move some logic around between components to make sure everything can be dynamic
- Keeps all the search functionality (switching product types etc.) on a single tab to prevent an additional API request.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Checkout this branch.
2. Build WooCommerce Admin on your local environment: `cd plugins/woocommerce-admin` then `nvm use && pnpm build`
3. Visit WooCommerce > Extensions (e.g. http://localhost:8888/wp-admin/admin.php?page=wc-admin&path=%2Fextensions&tab=extensions)
4. Perform a search for `store` – this will give some good results to compare
5. See that the page loads with a search tab and that there are “view all” buttons showing below both Extensions and Themes results
6. See that no category selector appears on the all results page
7. Click on either of the “View All” buttons and see that the results for that type are loaded correctly
8. See that the correct category selector appears
9. See that you should also be able to navigate back using browser history
10. See that after clicking the “View All” button for extensions, you can navigate to the “Themes” tab directly (previously this would display the incorrect product images)
11. Test around this issue

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

</details>

kdevnel committed:
Marketplace: consolidate extension/theme component

kdevnel committed:
Fix bug with tab routing when category is set

kdevnel committed:
Add View All button functionality

- make category selector visibility dynamic
- add search result components and matching tabs
- add button to direct to relevant result type

kdevnel committed:
Add styling to buttons

kdevnel commented on plugins/woocommerce-admin/client/marketplace/components/search-results/search-results.tsx:
```diff
@@ -19,15 +34,46 @@ export default function SearchResults( props: SearchResultProps ): JSX.Element {
+				) }
+				variant="secondary"
+				text="View all"
+				onClick={ () => navigateToTab( ProductType.extension ) }
```
Since this is a button, I think this would have ideally been an `href` property instead of an `onClick` but this is consistent with the tabs and category selector behaviour and prevents a full page reload.

andfinally replied:
TIL why React UIs do this! It's always bugged me when I can't right click and do "Open Link in New Tab". But not reloading the page is probably more valuable.

bgrgicak commented on plugins/woocommerce-admin/client/marketplace/components/extensions/extensions.tsx:
```diff
@@ -17,60 +18,75 @@ import { Product, ProductType } from '../product-list/types';
+	labelPlural: string;
 }
 
 export default function Extensions( props: ExtensionsProps ): JSX.Element {
```
Could we rename this to something like `Products`?

kdevnel replied:
Done in 6263b0025b807d0d3ea3089b0cb774bad59f61d2

bgrgicak commented on plugins/woocommerce-admin/client/marketplace/components/extensions/extensions.tsx:
```diff
@@ -17,60 +18,75 @@ import { Product, ProductType } from '../product-list/types';
 	const marketplaceContextValue = useContext( MarketplaceContext );
 	const { isLoading } = marketplaceContextValue;
+	const type =
+		props.label === 'extension' ? ProductType.extension : ProductType.theme;
```
This looks like it could lead to issues.
Could we flip it? We can have `type` as a prop and define the labels inside the component.
This would also be consistent with some other components where the type is passed in to modify the component behavior.

kdevnel replied:
Yes, this makes way more sense! Will work on that for my next commit.

kdevnel replied:
Done in 6263b0025b807d0d3ea3089b0cb774bad59f61d2

bgrgicak commented on plugins/woocommerce-admin/client/marketplace/components/search-results/search-extensions.tsx:
```diff
@@ -0,0 +1,30 @@
+import './search-results.scss';
+import { Product, ProductType } from '../product-list/types';
+import Extensions from '../extensions/extensions';
+//import { MARKETPLACE_SEARCH_RESULTS_PER_PAGE } from '../constants';
```
Can we remove this line?

bgrgicak commented on plugins/woocommerce-admin/client/marketplace/components/content/content.tsx:
```diff
@@ -104,11 +105,29 @@ export default function Content(): JSX.Element {
 			case 'search':
 				return <SearchResults products={ products } />;
+			case 'search-extensions':
+				return <SearchExtensions products={ products } />;
+			case 'search-themes':
+				return <SearchThemes products={ products } />;
```
We could use `SearchResults` with a `type` property. The current implementation adds extra components that don't do a lot and could easily be merged into `SearchResults`.

kdevnel replied:
Done in 6263b0025b807d0d3ea3089b0cb774bad59f61d2

bgrgicak commented on plugins/woocommerce-admin/client/marketplace/components/search-results/search-results.tsx:
```diff
@@ -19,15 +34,46 @@ export default function SearchResults( props: SearchResultProps ): JSX.Element {
+		navigateTo( {
+			url: getNewPath( {
+				category: ALL_CATEGORIES_SLUGS[ tab ],
+				tab: TAB[ tab ],
```
Not sure if this is a good idea, but what if we keep the tab as search, but add another prop like `section`.
This would simplify the code in `Tabs` and remove the need for having `ALL_CATEGORIES_SLUGS`.

The `section` could just show or hide parts of `SearchResults`.

kdevnel replied:
This is definitely a better approach, though I think both could work. I've refactored to do things this way in 6263b0025b807d0d3ea3089b0cb774bad59f61d2

kdevnel committed:
Handle review feedback and consolidate compoonents

kdevnel committed:
Merge branch 'update/in-app-multiple-category-filters' into add/wccom-18213-marketplace-view-more-button

kdevnel committed:
Show button at the right time and add translation

andfinally reviewed (commented):
The code looks good – nice refactor! I'm seeing a few wrinkles in the search behaviour we'll need to iron out. I've tried to split out the ones that make more sense in a separate issue, to 18332-gh-Automattic/woocommerce.com. Let me know if you think we should shift any of these to that issue.

If I do a search, go to the "all results" themes view, and then filter by a category that has no results, I get `0 themes found`, but no indication that the category is selected. We should probably continue to show categories even in this situation.

https://github.com/woocommerce/woocommerce/assets/1647564/78b4a7d9-c536-4f73-9931-8264a35332ed

The "View all" button appears on the `Browse` screen, but not the `Themes` screen. Is that expected? I guess we're showing it when there are more results than we can show on that tab.

<img width="500" src="https://github.com/woocommerce/woocommerce/assets/1647564/33bf732b-050a-4868-bd31-7a8d1d30ac9c">

When there are no extensions results, there is also not enough margin between the "Most Popular Extensions" element and the themes results.

![image](https://github.com/woocommerce/woocommerce/assets/1647564/1165da48-6843-442d-a5db-30ca3386c529)

kdevnel commented:
Thanks @andfinally – that's great feedback. I think I will keep those 3 items you mentioned in this PR and commit some fixes for them today.

kdevnel committed:
Address PR feedback

- ensure categories show when there are no results on all product type search
- fix spacing between extensions and themes when no results available
- ensure view all button only shows on search results page

bgrgicak committed:
Merge branch 'feature/marketplace-themes' into add/wccom-18213-marketplace-view-more-button

bgrgicak commented on plugins/woocommerce-admin/client/marketplace/components/products/products.tsx:
```diff
@@ -0,0 +1,131 @@
+	[ ProductType.theme ]: {
+		label: 'themes',
+		singularLabel: 'theme',
+	},
```
These need to be translatable.

andfinally replied:
Addressed in [`9560c0f` (#40375)](https://github.com/woocommerce/woocommerce/pull/40375/commits/9560c0faf5859aa9bf0b4942bea5dee5b78dc820).

bgrgicak commented on plugins/woocommerce-admin/client/marketplace/components/products/products.tsx:
```diff
@@ -0,0 +1,131 @@
+	if ( productTotalCount > 0 ) {
+		title = sprintf(
+			// translators: %1$s: number of items, %2$s: singular item label, %3$s: plural item label
+			_n( '%1$s %2$s', '%1$s %3$s', productTotalCount, 'woocommerce' ),
```
Does this need to be translatable?

andfinally committed:
Addressing feedback. Translating labels.

bgrgicak commented on plugins/woocommerce-admin/client/marketplace/components/products/products.tsx:
```diff
@@ -0,0 +1,131 @@
+
+const LABELS = {
+	[ ProductType.extension ]: {
+		label: __( 'extensions' ),
+		singularLabel: __( 'extension' ),
+	},
+	[ ProductType.theme ]: {
+		label: __( 'themes' ),
+		singularLabel: __( 'theme' ),
```
```suggestion
		label: __( 'extensions', 'woocommerce' ),
		singularLabel: __( 'extension', 'woocommerce' ),
	},
	[ ProductType.theme ]: {
		label: __( 'themes', 'woocommerce' ),
		singularLabel: __( 'theme', 'woocommerce' ),
```

bgrgicak committed:
Update plugins/woocommerce-admin/client/marketplace/components/products/products.tsx

andfinally reviewed (approved):
Thanks for the tweaks! This works for me.
